feat: add third-party auth docs (#28423)
supabase/supabase: apps/docs/content/guides/auth/third-party/firebase-auth.mdx

---
id: 'auth-third-party-firebase-auth'
title: 'Firebase Auth'
subtitle: 'Use Firebase Auth with your Supabase project'
---

<Admonition type="note">

Creating a Firebase Auth third-party integration with your Supabase project is currently limited to a private alpha release. We are still trying to improve the developer experience around securing your project. Please reach out to [Supabase Support](https://supabase.help) if you're interested in having this enabled on your project.

</Admonition>

Firebase Auth can be used as a third-party authentication provider alongside Supabase Auth, or standalone, with your Supabase project.

## Getting started

1. First you need to add an integration to connect your Supabase project with your Firebase project. You will need to get the Project ID in the [Firebase Console](https://console.firebase.google.com/u/0/project/_/settings/general).
2. Add a new Third-party Auth integration in your project's [Authentication settings](/dashboard/project/_/settings/auth).

> 3. Create and attach restrictive RLS policies to all tables in your public schema, Storage and Realtime to prevent unauthorized access from unrelated Firebase projects.

Reviewer: ```suggestion
3. If you are using Third Party Auth in local development, create and attach restrictive RLS policies to all tables in your public schema, Storage and Realtime to prevent unauthorized access from unrelated Firebase projects.
```

4. Assign the `role: 'authenticated'` [custom user claim](https://firebase.google.com/docs/auth/admin/custom-claims) to all your users.
5. Finally set up the Supabase client in your application.

<Tabs type="underlined" queryGroup="firebase-create-client">

<TabPanel id="create-client-web" label="TypeScript">

Creating a client for the Web is as easy as passing the `accessToken` async function. This function should [return the Firebase Auth JWT of the current user](https://firebase.google.com/docs/auth/admin/verify-id-tokens#web) (or null if no such user) is found.

```typescript
import { createClient } from '@supabase/supabase-js'

const supabase = createClient('https://<supabase-project>.supabase.co', 'SUPABASE_ANON_KEY', {
  accessToken: async () => {
    ;(await firebase.auth().currentUser?.getIdToken(/* forceRefresh */ false)) ?? null
  },
})
```

> Make sure the all users in your application have the `role: 'authenticated'` [custom claim](https://firebase.google.com/docs/auth/admin/custom-claims) set. If you're using the `onCreate` Cloud Function to add this custom role to newly signed up users, you will need to call `getIdToken(/* forceRefresh */ true)` immediately after sign up as the `onCreate` function does not run synchronously.

Reviewer: since we already go over this below, i think we are repeating the same information here (oncreate not being synchronous etc). We could create a new heading called "Setting up the supabase client" and move this example there. 

That way getting started just contains the 5 bullet points, and it links to the 3 sections below
- Assign the "role" custom claim
- Setting up the supabase client
- Securing your project with restrictive RLS policies - This section could be moved all the way down since it is relevant for self hosting users only.

</TabPanel>

<TabPanel id="create-client-swift" label="Swift (iOS)">

TODO!

</TabPanel>

</Tabs>

## Securing your project with restrictive RLS policies

<Admonition type="note">

> We are actively working to reduce the burden placed on developers to properly secure access to their project's data when using Firebase Auth. If you are self-hosting or using this in the private alpha stage, make sure you closely follow these instructions.

Reviewer: ```suggestion
If you are self-hosting, make sure you closely follow these instructions. If you are using the Firebase provider in the hosted platform, requests from other Firebase projects are rejected with a 401 status code. 
```

        and
        auth.jwt()->>'aud' = '<firebase-project-id>'
     )
  );
```

If you have a lot of tables in your app, or need to manage complex RLS policies for [Storage]() or [Realtime]() it can be useful to define a [stable Postgres function](https://www.postgresql.org/docs/current/xfunc-volatility.html) that performs the check to cut down on duplicate code. For example:

```sql
create function public.is_supabase_or_firebase_project_jwt()
  returns bool
  language sql
  stable
  returns null on null input
  return (
    (auth.jwt()->>'iss' = 'https://<project-ref>.supabase.co/auth/v1')
    or
    (
        auth.jwt()->>'iss' = concat('https://securetoken.google.com/<firebase-project-id>')
        and
        auth.jwt()->>'aud' = '<firebase-project-id>'
     )
  );
```

Please make sure you substitute `<project-ref>` with your Supabase project's ID and the `<firebase-project-id>` to your Firebase Project ID. Then the restrictive policies on all your tables, buckets and channels can be simplified to be:

```sql
create policy "Restrict access to correct Supabase and Firebase projects"
  on table_name
  as restrictive
  to authenticated

>   using (public.is_supabase_or_firebase_project_jwt() is true);

Reviewer: should the function call be prefixed with select for performance? https://supabase.com/docs/guides/database/database-advisors?queryGroups=lint&lint=0003_auth_rls_initplan

Author: Huh! Neat trick! Didn't know about this. It should, yes!

```

## Assign the "role" custom claim

Your Supabase project inspects the `role` claim present in all JWTs sent to it, to assign the correct Postgres role when using the Data API, Storage or Realtime authorization.

By default, Firebase JWTs do not contain a `role` claim in them. If you were to send such a JWT to your Supabase project, the `anon` role would be assigned when executing the Postgres query. Most of your app's logic will be accessible by the `authenticated` role.

> The easiest way to achieve this is to define a [blocking Firebase Authentication function](https://firebase.google.com/docs/auth/extend-with-blocking-functions). However this is only available if your Firebase project uses [Firebase Authentication with Identity Platform](https://cloud.google.com/identity-platform/docs/product-comparison).

Reviewer: we could link to the sub-sections below instead of linking to the firebase docs here. to make it very explicit that there are 2 ways and they are described in the sections below. we can link to the firebase docs from the subsections instead so that the flow is clear

        and
        auth.jwt()->>'aud' = '<firebase-project-id>'
     )
  );
```

Make sure you substitute `<project-ref>` with your Supabase project's ID and the `<firebase-project-id>` to your Firebase Project ID. Then the restrictive policies on all your tables, buckets and channels can be simplified to be:

```sql
create policy "Restrict access to correct Supabase and Firebase projects"
  on table_name
  as restrictive
  to authenticated
  using ((select public.is_supabase_or_firebase_project_jwt()) is true);
```

## Assign the "role" custom claim

Your Supabase project inspects the `role` claim present in all JWTs sent to it, to assign the correct Postgres role when using the Data API, Storage or Realtime authorization.

By default, Firebase JWTs do not contain a `role` claim in them. If you were to send such a JWT to your Supabase project, the `anon` role would be assigned when executing the Postgres query. Most of your app's logic will be accessible by the `authenticated` role.

### Use Firebase Authentication functions to assign the authenticated role

You have two choices to set up a Firebase Authentication function depending on your Firebase project's configuration:

> 1. Easiest: Use a [blocking Firebase Authentication function](https://firebase.google.com/docs/auth/extend-with-blocking-functions) but this is only available if your project uses [Firebase Authentication with Identity Platform]().

Reviewer 2: Not sure what the intended link is here this one was the best fit I could find from quick search
```suggestion
1. Easiest: Use a [blocking Firebase Authentication function](https://firebase.google.com/docs/auth/extend-with-blocking-functions) but this is only available if your project uses [Firebase Authentication with Identity Platform](https://cloud.google.com/security/products/identity-platform).
```
